packages/react-core/src/components/DualListSelector/examples/ComposableTree.tsx

```diff
@@ -0,0 +1,344 @@
+      const otherCheckedNodeNames = prevChecked.filter(id => !nodeIdsToCheck.includes(id));
+      return !isChecked ? otherCheckedNodeNames : [...otherCheckedNodeNames, ...nodeIdsToCheck];
+    });
+  };
+
+  // builds a search input - used in each dual list selector pane
+  const buildSearchInput = (isChosen: boolean) => {
+    const onChange = value => {
+      if (isChosen) {
+        hiddenChosen = [];
+      } else {
+        hiddenAvailable = [];
+      }
+      return isChosen ? setChosenFilter(value) : setAvailableFilter(value);
+    };
+
+    return (
+      <SearchInput value={isChosen ? chosenFilter : availableFilter} onChange={onChange} onClear={() => onChange('')} />
+    );
+  };
+
+  // Builds the DualListSelectorTreeItems from the FoodNodes
+  const buildOptions = (
+    isChosen: boolean,
+    [node, ...remainingNodes]: FoodNode[],
+    hasParentMatch: boolean
+  ): DualListSelectorTreeItemData[] => {
+    if (!node) {
+      return [];
+    }
+
+    const isChecked = isNodeChecked(node, isChosen);
+
+    const filterValue = isChosen ? chosenFilter : availableFilter;
+    const descendentLeafIds = memoizedLeavesById[node.id];
+    const descendentsOnThisPane = isChosen
+      ? descendentLeafIds.filter(id => chosenLeafIds.includes(id))
+      : descendentLeafIds.filter(id => !chosenLeafIds.includes(id));
+
+    const hasMatchingChildren =
+      filterValue && descendentsOnThisPane.some(id => memoizedNodeText[id].includes(filterValue));
+    const isFilterMatch = filterValue && node.text.includes(filterValue) && descendentsOnThisPane.length > 0;
+
+    // A node is displayed if:
+    //   - There is no filter value and this node or its descendents belong on this pane
+    //   - There is a filter value and this node or one of this node's descendents or ancestors match on this pane
+    const isDisplayed =
+      (!filterValue && descendentsOnThisPane.length > 0) || hasMatchingChildren || hasParentMatch || isFilterMatch;
+
+    if (!isDisplayed) {
+      if (isChosen) {
+        hiddenChosen.push(node.id);
+      } else {
+        hiddenAvailable.push(node.id);
+      }
+    }
+
+    return [
+      ...(isDisplayed
+        ? [
+            {
+              id: node.id,
+              text: node.text,
+              isChecked,
+              checkProps: { 'aria-label': `Select ${node.text}` },
+              hasBadge: node.children && node.children.length > 0,
+              badgeProps: { isRead: true },
+              defaultExpanded: isChosen ? !!chosenFilter : !!availableFilter,
+              children: node.children
+                ? buildOptions(isChosen, node.children, isFilterMatch || hasParentMatch)
+                : undefined
+            }
+          ]
+        : []),
+      ...(!isDisplayed && node.children && node.children.length
+        ? buildOptions(isChosen, node.children, hasParentMatch)
+        : []),
+      ...(remainingNodes ? buildOptions(isChosen, remainingNodes, hasParentMatch) : [])
+    ];
```

Comment:
This is pretty dense, I think some newlines between the spreads would improve readability if nothing else.